src/mongocrypt-ctx-rewrap-many-datakey.c

```diff
@@ -0,0 +1,366 @@
+
+   BSON_ASSERT_PARAM (ctx);
+
+   /* Finish KMS requests for decryption if there were any. */
+   if (ctx->state == MONGOCRYPT_CTX_NEED_KMS) {
+      _mongocrypt_opts_kms_providers_t *const providers =
+         _mongocrypt_ctx_kms_providers (ctx);
+
+      if (!_mongocrypt_key_broker_kms_done (&ctx->kb, providers)) {
+         _mongocrypt_status_copy_to (ctx->kb.status, ctx->status);
+         return _mongocrypt_ctx_fail (ctx);
+      }
+   }
+
+   /* For all decrypted datakeys, initialize a corresponding datakey context. */
+   {
+      key_returned_t *key;
+
+      /* Some decrypted keys may have been cached. */
+      for (key = ctx->kb.keys_cached; key; key = key->next) {
+         if (!_add_new_datakey (ctx, key)) {
+            return _mongocrypt_ctx_fail (ctx);
+         }
+      }
+
+      /* Remaining keys were decrypted via KMS requests. */
+      for (key = ctx->kb.keys_returned; key; key = key->next) {
+         if (!_add_new_datakey (ctx, key)) {
+            return _mongocrypt_ctx_fail (ctx);
+         }
+      }
+   }
+
+   /* Prepare iterator used by _next_kms_ctx_encrypt. */
+   rmdctx->datakeys_iter = rmdctx->datakeys;
+
+   /* Skip datakeys that do not require a KMS request. */
+   while (rmdctx->datakeys_iter &&
+          rmdctx->datakeys_iter->dkctx->state == MONGOCRYPT_CTX_READY) {
+      rmdctx->datakeys_iter = rmdctx->datakeys_iter->next;
+   }
+
+   /* Skip to READY state if no KMS requests are required. */
+   if (!rmdctx->datakeys_iter) {
+      ctx->state = MONGOCRYPT_CTX_READY;
+      return true;
+   }
+
+   ctx->state = MONGOCRYPT_CTX_NEED_KMS;
+   ctx->vtable.next_kms_ctx = _next_kms_ctx_encrypt;
+   ctx->vtable.kms_done = _kms_done_encrypt;
+
+   return true;
+}
+
+static bool
+_mongo_done_keys (mongocrypt_ctx_t *ctx)
+{
+   BSON_ASSERT_PARAM (ctx);
+
+   if (!_mongocrypt_key_broker_docs_done (&ctx->kb) ||
+       !_mongocrypt_ctx_state_from_key_broker (ctx)) {
+      return _mongocrypt_ctx_fail (ctx);
+   }
+
+   /* No keys to rewrap, no work to be done. */
+   if (!ctx->kb.key_requests) {
+      ctx->state = MONGOCRYPT_CTX_DONE;
+      return true;
+   }
+
+   /* No KMS required, skip straight to encryption. */
+   if (ctx->state == MONGOCRYPT_CTX_READY) {
+      return _start_kms_encrypt (ctx);
+   }
+
+   /* KMS requests needed to decrypt keys. */
+   BSON_ASSERT (ctx->state == MONGOCRYPT_CTX_NEED_KMS);
+
+   return true;
+}
+
+static bool
+_mongo_op_keys (mongocrypt_ctx_t *ctx, mongocrypt_binary_t *out)
+{
+   _mongocrypt_ctx_rewrap_many_datakey_t *const rmdctx =
+      (_mongocrypt_ctx_rewrap_many_datakey_t *) ctx;
+
+   BSON_ASSERT_PARAM (ctx);
+   BSON_ASSERT_PARAM (out);
+
+   _mongocrypt_buffer_to_binary (&rmdctx->filter, out);
+
+   return true;
+}
+
+static bool
+_kms_start_decrypt (mongocrypt_ctx_t *ctx)
+{
+   BSON_ASSERT_PARAM (ctx);
+
+   return _mongocrypt_key_broker_request_any (&ctx->kb) &&
+          _mongocrypt_key_broker_requests_done (&ctx->kb) &&
```

Comment:
It would be simpler. I introduced `KB_REQUESTING_ANY` so that the interface preserved the existing pattern of calling `_mongocrypt_key_broker_requests_done` to end the `KB_REQUESTING` state. If we're fine with going straight to `_mongocrypt_key_broker_add_doc` calls after invoking `_mongocrypt_key_broker_request_any`, then we can remove the `KB_REQUESTING_ANY` state.

Reply:
I have removed the `KB_REQUESTING_ANY` state. Please let me know if it looks fine, or if you would prefer the prior interface.

Reply:
The removal the `KB_REQUESTING_ANY` state LGTM.